core/test/base/array.cpp

```diff
@@ -591,4 +592,106 @@ TYPED_TEST(Array, MoveArrayToView)
+    ASSERT_EQ(exec, this->x.get_executor());
+    ASSERT_EQ(exec, view.get_executor());
+    ASSERT_TRUE(this->x.is_owning());
+    ASSERT_FALSE(view.is_owning());
+}
+
+
+TYPED_TEST(Array, AsConstView)
+{
+    auto ptr = this->x.get_data();
+    auto size = this->x.get_num_elems();
+    auto exec = this->x.get_executor();
+    auto view = this->x.as_const_view();
+
+    ASSERT_EQ(ptr, this->x.get_data());
+    ASSERT_EQ(ptr, view.get_data());
+    ASSERT_EQ(size, this->x.get_num_elems());
+    ASSERT_EQ(size, view.get_num_elems());
+    ASSERT_EQ(exec, this->x.get_executor());
+    ASSERT_EQ(exec, view.get_executor());
+    ASSERT_TRUE(this->x.is_owning());
+    ASSERT_FALSE(view.is_owning());
+}
+
+
+TYPED_TEST(Array, ArrayConstCastWorksOnView)
+{
+    auto ptr = this->x.get_data();
+    auto size = this->x.get_num_elems();
+    auto exec = this->x.get_executor();
+    auto const_view = this->x.as_const_view();
+    auto view = gko::detail::array_const_cast(std::move(const_view));
+    static_assert(std::is_same<decltype(view), decltype(this->x)>::value,
+                  "wrong return type");
+
+    ASSERT_EQ(nullptr, const_view.get_data());
+    ASSERT_EQ(0, const_view.get_num_elems());
+    ASSERT_EQ(exec, const_view.get_executor());
+    ASSERT_EQ(ptr, this->x.get_data());
+    ASSERT_EQ(ptr, view.get_data());
+    ASSERT_EQ(size, this->x.get_num_elems());
+    ASSERT_EQ(size, view.get_num_elems());
+    ASSERT_EQ(exec, this->x.get_executor());
+    ASSERT_EQ(exec, view.get_executor());
+    ASSERT_TRUE(this->x.is_owning());
+    ASSERT_FALSE(view.is_owning());
+}
+
+
+TYPED_TEST(Array, ArrayConstCastWrapsCustomDeleter)
+{
+    auto data = this->exec->template alloc<TypeParam>(10);
+    bool deleted = false;
+
+    {
+        gko::Array<const TypeParam> array(
+            this->exec, 10, data, [&](const TypeParam* ptr) {
+                if (ptr == data && !deleted) {
+                    deleted = true;
+                    this->exec->free(const_cast<TypeParam*>(ptr));
+                } else {
+                    FAIL();
+                }
+            });
+        ASSERT_FALSE(array.is_owning());
+        auto mutable_array = gko::detail::array_const_cast(std::move(array));
+
+        ASSERT_EQ(nullptr, array.get_data());
+        ASSERT_EQ(0, array.get_num_elems());
+        ASSERT_EQ(this->exec, array.get_executor());
+        ASSERT_EQ(data, mutable_array.get_data());
+        ASSERT_EQ(10, mutable_array.get_num_elems());
+        ASSERT_EQ(this->exec, mutable_array.get_executor());
+        ASSERT_FALSE(mutable_array.is_owning());
+    }
+    ASSERT_TRUE(deleted);
+}
+
+
+TYPED_TEST(Array, ArrayConstCastWrapsDefaultDeleter)
+{
+    auto data = this->exec->template alloc<TypeParam>(10);
+    gko::Array<const TypeParam> array(this->exec, 10, data);
+    ASSERT_TRUE(array.is_owning());
+
+    auto mutable_array = gko::detail::array_const_cast(std::move(array));
+
+    ASSERT_TRUE(mutable_array.is_owning());
+}
```

Comment:
If we assume an const array doesn't own data, this test needs to be removed.
(See below)
```suggestion
```